sequencer/src/state.rs

```diff
@@ -308,6 +321,228 @@ fn validate_and_apply_header(
+                fee_merkle_tree.height(),
+            );
+
+            storage
+                .write()
+                .await
+                .store_state::<FeeMerkleTree, 256>(proof.proof, path, block_number)
+                .await
+                .context("failed to insert merkle nodes for block merkle tree")?;
+        }
+
+        if block_number == 0 {
+            continue;
+        }
+
+        // Insert block merkle tree nodes
+        let (_, proof) = block_merkle_tree
+            .lookup(block_number - 1)
+            .expect_ok()
+            .context("Index not found in block merkle tree")?;
+        let path = <u64 as ToTraversalPath<3>>::to_traversal_path(
+            &(block_number - 1),
+            block_merkle_tree.height(),
+        );
+
+        storage
+            .write()
+            .await
+            .store_state::<BlockMerkleTree, 3>(proof.proof, path, block_number)
+            .await
+            .context("failed to insert merkle nodes for block merkle tree")?;
+
+        storage
+            .write()
+            .await
+            .set_last_state_height(block_number as usize)
+            .await?;
+    }
+
+    Ok(())
+}
+
+impl ValidatedState {
+    async fn apply_header(
+        &self,
+        instance: &NodeState,
+        parent_leaf: &Leaf,
+        proposed_header: &Header,
+        catchup: impl StateCatchup,
+    ) -> anyhow::Result<(Self, Delta)> {
+        // Clone state to avoid mutation. Consumer can take update
+        // through returned value.
+
+        let mut validated_state = self.clone();
+
+        // Fetch the new L1 deposits between parent and current finalized L1 block.
+        let l1_deposits = if let Some(block_info) = proposed_header.l1_finalized {
+            instance
+                .l1_client
+                .get_finalized_deposits(
+                    parent_leaf
+                        .get_block_header()
+                        .l1_finalized
+                        .map(|block_info| block_info.number),
+                    block_info.number,
+                )
+                .await
+        } else {
+            vec![]
+        };
+
+        let mut delta = Delta::default();
+
+        catchup
+            .remember_blocks_merkle_tree(
+                parent_leaf.get_view_number(),
+                &mut validated_state.block_merkle_tree,
+            )
+            .await?;
+
+        for FeeInfo { account, .. } in l1_deposits {
+            let account_proof = catchup
+                .fetch_accounts(
+                    parent_leaf.get_view_number(),
+                    validated_state.fee_merkle_tree.commitment(),
+                    vec![account],
+                )
+                .await?;
+
+            let proof = &account_proof.first().unwrap().proof;
+
+            proof
+                .remember(&mut validated_state.fee_merkle_tree)
+                .unwrap();
+            delta.fees_delta.insert(account);
+        }
+
```

Comment:
Missing stuff from `validate_and_apply_header`, e.g. proposer fee and block merkle tree